### PR TITLE
Improve NixOS module

### PR DESCRIPTION
**Pull request by erikarvstedt**

### CODE REVIEW

**Comment by erikarvstedt** [comment]
Rebased.

**Comment by lovesegfault** [comment]
Pinging @rycee for review

File: nixos/default.nix [hunk 1]
```diff
@@ -45,6 +46,11 @@ in {
         <option>users.users.‹name?›.packages</option> option.
       '';
 
+      useGlobalPkgs = mkEnableOption ''
+        using <literal>pkgs</literal> from the NixOS modules in Home Manager.
```
Comment (cole-h) [inline nixos/default.nix]:
```suggestion
        use <literal>pkgs</literal> from the NixOS modules in Home Manager.
```

File: modules/misc/news.nix [hunk 2]
```diff
@@ -1333,13 +1333,23 @@ in
...
           A new module is available: 'programs.abook'
         '';
       }
+
+      {
+        time = "2020-03-05T00:00:00+00:00";
+        condition = hostPlatform.isLinux;
+        message = ''
+          The NixOS integration module has a new option: 'home-manager.useGlobalPkgs'.
+          To learn more, see the installation section of the manual.
```
Comment (cole-h) [inline modules/misc/news.nix]: Any chance of a small-ish description, other than "see the manual"?

File: README.md [hunk 3]
```diff
@@ -6,8 +6,8 @@ using the [Nix][] package manager together with the Nix libraries
 found in [Nixpkgs][]. Before attempting to use Home Manager please
 read the warning below.
 
-For a more systematic overview of all the options Home Manager
-provides please see the [Home Manager manual][configuration options].
+For a more systematic documentation of Home Manager and all available
+options please see the [Home Manager manual][manual].
```
[the remark this answers is not in the document]
Comment (erikarvstedt) [reply README.md]: Fixed. Editing help from native speakers is highly welcome, thank you!

File: doc/installation.xml [hunk 4]
```diff
@@ -225,6 +225,44 @@ home-manager.useUserPackages = true;
     become the default value in the future.
    </para>
   </note>
+
+  <note>
+   <para>
+    By default, Home Manager modules use a private <literal>pkgs</literal> instance
+    that's configured via the <option>nixpkgs</option> options.
```
[the remark this answers is not in the document]
Comment (erikarvstedt) [reply doc/installation.xml]: Fixed.

File: doc/installation.xml [hunk 5]
```diff
@@ -225,6 +225,44 @@ home-manager.useUserPackages = true;
...
+   This saves an extra nixpkgs evaluation, adds consistency, and removes the
+   dependency on <literal>NIX_PATH</literal>, which is otherwise used for
+   importing the pkgs.
+   This option may become the default value in the future.
+   </para>
+  </note>
+
+  <para>
+   To easily set Home Manager options for a specific user, use the following recipe.
+   This is useful when a user config is composed through multiple NixOS modules.
```
[the remark this answers is not in the document]
Comment (erikarvstedt) [reply doc/installation.xml]: Fixed.

File: nixos/default.nix (hunk 1 above)
Comment (erikarvstedt) [reply nixos/default.nix]: `mkEnableOption` prefixes the description with 'Whether to enable'. The definition of
`useUserPackages` uses a similar style of accomodating to that prefix.

File: modules/misc/news.nix (hunk 2 above)
Comment (erikarvstedt) [reply modules/misc/news.nix]: Yes, fixed.

File: modules/modules.nix [hunk 6]
```diff
@@ -170,11 +173,14 @@ let
...
   pkgsModule = {
-    config._module.args.baseModules = modules;
-    config._module.args.pkgs = lib.mkDefault pkgs;
-    config._module.check = check;
-    config.lib = lib.hm;
-    config.nixpkgs.system = mkDefault pkgs.system;
+    config = {
+      _module.args.baseModules = modules;
+      _module.args.pkgs = lib.mkDefault pkgs;
+      _module.check = check;
```
[the remark this answers is not in the document]
Comment (erikarvstedt) [reply modules/modules.nix]: ∞ recursion would only happen in a `rec` attrset.
I wanted to keep the change minimal. I'll let @rycee decide whether to add a `_module` grouping. (I think `args` can stay inline.)

File: nixos/default.nix (hunk 1 above)
Comment (cole-h) [reply nixos/default.nix]: Ah, I missed that. Thanks for the clarification!

File: modules/modules.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (cole-h) [reply modules/modules.nix]: Fair enough. (I'm not terribly familiar with Nix yet, but just enough to have been bitten by infinite recursion, which is why that was a concern :P)

File: nixos/default.nix (hunk 1 above)
Comment (rycee) [reply nixos/default.nix]: I think this is fine as-is. But in general, a useful idiom when the default description format of `mkEnableOption` is insufficient is to use something like
```nix
mkEnableOption "" // {
  description = ''
    A custom description goes here.
  '';
}
```

Comment (rycee) [inline nixos/default.nix]: Nothing important but perhaps 
```suggestion
        using the system configuration's <literal>pkgs</literal> argument in Home Manager.
```
or something would be more clear?

File: modules/modules.nix [hunk 7]
```diff
@@ -5,6 +5,9 @@
 
   # Whether to enable module type checking.
 , check ? true
+
+# If disabled, the pkgs passed to this function are used instead
```
Comment (rycee) [inline modules/modules.nix]:
```suggestion
# If disabled, the pkgs passed to this function is used instead.
```

File: modules/modules.nix [hunk 8]
```diff
@@ -170,11 +173,14 @@ let
...
-    config._module.args.pkgs = lib.mkDefault pkgs;
-    config._module.check = check;
-    config.lib = lib.hm;
-    config.nixpkgs.system = mkDefault pkgs.system;
+    config = {
+      _module.args.baseModules = modules;
+      _module.args.pkgs = lib.mkDefault pkgs;
+      _module.check = check;
+      lib = lib.hm;
+    } // optionalAttrs (useNixpkgsModule) {
```
Comment (rycee) [inline modules/modules.nix]:
```suggestion
    } // optionalAttrs useNixpkgsModule {
```

File: modules/modules.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (rycee) [reply modules/modules.nix]: I think this is OK as-is.

**Comment by rycee** [comment]
I cherry picked the `README: add link to main manual page` commit to master in c7b43786ad5740f8ae054cb97d3493c585920031.

File: doc/installation.xml [hunk 9]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
     become the default value in the future.
    </para>
   </note>
+
+  <note>
+   <para>
+    By default, Home Manager modules use a private <literal>pkgs</literal> instance
+    that is configured via the <option>nixpkgs</option> options.
```
Comment (rycee) [inline doc/installation.xml]:
```suggestion xml
    that is configured via the <option>home-manager.users.&lt;name?&gt;.nixpkgs</option> options.
```

File: doc/installation.xml [hunk 10]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
...
+    To disable these options and to use the global <literal>pkgs</literal> from
+    the NixOS modules, set
```
Comment (rycee) [inline doc/installation.xml]:
```suggestion
    To disable these options and to use the global <literal>pkgs</literal>
    that is configured via the <option>nixpkgs</option> options, set
```

File: doc/installation.xml [hunk 11]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
...
+   </para>
+<programlisting language="nix">
+home-manager.useGlobalPkgs = true;
+</programlisting>
+  <para>
+   This saves an extra nixpkgs evaluation, adds consistency, and removes the
```
Comment (rycee) [inline doc/installation.xml]:
```suggestion
   This saves an extra Nixpkgs evaluation, adds consistency, and removes the
```

File: doc/installation.xml [hunk 12]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
...
+   dependency on <literal>NIX_PATH</literal>, which is otherwise used for
```
Comment (rycee) [inline doc/installation.xml]:
```suggestion xml
   dependency on <envar>NIX_PATH</envar>, which is otherwise used for
```

File: doc/installation.xml [hunk 13]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
...
+   importing the pkgs.
+   This option may become the default value in the future.
```
Comment (rycee) [inline doc/installation.xml]:
```suggestion
   This saves an extra Nixpkgs evaluation, adds consistency, and removes the
   dependency on <envvar>NIX_PATH</envvar>, which is otherwise used for
   importing Nixpkgs.
```

Comment (erikarvstedt) [reply doc/installation.xml]: No, `envar` from your deleted comment is actually correct :smiley_cat:

File: doc/installation.xml [hunk 14]
```diff
@@ -225,6 +225,45 @@ home-manager.useUserPackages = true;
...
+</programlisting>
+  <para>
+   This saves an extra Nixpkgs evaluation, adds consistency, and removes the
+   dependency on <envar>NIX_PATH</envar>, which is otherwise used for
+   importing the pkgs.
+   This option may become the default value in the future.
+   </para>
+  </note>
+
+  <para>
```
Comment (rycee) [inline doc/installation.xml]: I would suggest removing this commit since it is orthogonal to the general PR and I think does not fit in the installation section. Maybe could go in a "tips and tricks" section (in a separate PR) or even on the [wiki](https://nixos.wiki/wiki/Home_Manager)?

**Comment by erikarvstedt** [comment]
[Here's a rebased version](https://github.com/erikarvstedt/home-manager/tree/improve-nixos-module-rebased) of this PR.

File: doc/installation.xml (hunk 13 above)
Comment (rycee) [reply doc/installation.xml]: I always forget which way it is 🙂

File: doc/installation.xml (hunk 14 above)
Comment (erikarvstedt) [reply doc/installation.xml]: I've removed it. Where would you add that tips and tricks section?

**Comment by rycee** [comment]
Thanks! Added to master in efbe1383e64585ebc0303f4703e7e582aaf6b933.

File: doc/installation.xml (hunk 14 above)
Comment (rycee) [reply doc/installation.xml]: I guess there should be a new chapter called something like "NixOS Integration", with a section "Tips and Tricks" where this would go into a subsection.

I think there you could also have more space to write text surrounding this use pattern since it is dealing with more advanced concepts such as defining your own options and using a function like `mkAliasDefinitions`.
